Implement analytics

mikehardy opened the pull request:
I want to be much more thorough about tagging but this is ready for review

Two main commits 

- one is all the infrastructure required (and most is convenience methods)
- two is peppering the code with tags 

This works as is I just want better and more complete tagging. If you install the analytics app on your phone and put this in your Anki build you'll see yourself show up as a real-time user though, which is nice.

mikehardy commented:
As I tag things differently it'll change (and the questions we can answer will be more relevant) - but here's the current sample assuming you've got credentials: https://analytics.google.com/analytics/web/#/report/defaultid/a125800786w184475040p181795472/

mikehardy commented:
Closer now. 
ACRA/Analytics uncaught exception handlers interact well.
opt-out is built in to the code but not exposed yet
basic tagging for screen views is built into DialogFragment and AnkiActivity hierarchy, as well as our main preferences

Need to do tagging of specific events that are interesting, and catch the rest of the Activity and AsyncTask and Preferences screens and it's beta quality

mikehardy commented:
@timrae I think this is ready for a full review and test run. I've implemented a popup if the opt in preference isn't found, after that it's in the general preferences area, and I've verified that toggling things works as expected so we shouldn't have a trust violation. In the future we'll set "dry run" for DEBUG builds but for now it's useful to have them debug hits send

The first commit in particular is the entire infrastructure implementation, the rest are tagging code or codacy

Curious to see what you think, it's my first analytics implementation.

mikehardy commented:
Oh, and the analytics app from google is fairly amazing, it'll show you in near-real-time (maybe 5s delay?) as you test.

mikehardy commented on AnkiDroid/src/main/res/values/10-preferences.xml:
```diff
@@ -42,6 +42,7 @@
     <string name="pref_cat_performance">Performance</string>
     <string name="pref_cat_workarounds">Workarounds</string>
     <string name="pref_cat_plugins">Plugins</string>
+    <string name="pref_cat_analytics">Feature Usage</string>
```
I don't think this is needed actually - should remove

mikehardy commented on AnkiDroid/src/main/java/com/ichi2/anki/DeckPicker.java:
```diff
@@ -853,6 +855,11 @@ public void addNote() {
+        if (!preferences.contains(UsageAnalytics.ANALYTICS_OPTIN_KEY)) {
+            showDialogFragment(DeckPickerAnalyticsOptInDialog.newInstance());
+        }
+
```
I didn't know the best place to trigger the first-time pop-up, I'm not sure if this is the preferred spot, but it does work

mikehardy replied:
I decided to keep it simple, I like your suggestion. I moved the popup to the upgrade block and will just leave it at that. In a way it'll filter the analytics to users that should have more say data-driven decision-wise, and also it's on me/us to generate an upgrade worthy of publish periodically to opt-in new users right? 

(here's some Spanish for your trip: "Òjala no haya diàlogo de corrupciòn y este diàlogo a la vez, que vergŭenza" - "god willing (super common phrase, go with it), there isn't a corruption dialog and this dialog at the same time, how embarrassing)

mikehardy commented on AnkiDroid/build.gradle:
```diff
@@ -107,6 +107,9 @@ task jacocoTestReport(type: JacocoReport, dependsOn: ['testDebugUnitTest', 'crea
 dependencies {
+    compileOnly "com.google.auto.service:auto-service:1.0-rc4"
+    annotationProcessor "com.google.auto.service:auto-service:1.0-rc4"
+
```
ACRA allows you to do extensions with code generated via annotation, these are the deps that do that

mikehardy commented on AnkiDroid/src/main/java/com/ichi2/anki/analytics/UsageAnalytics.java:
```diff
@@ -0,0 +1,205 @@
+     * @param label A label for the event, may be null
+     */
+    public static void sendAnalyticsEvent(@NonNull String category, @NonNull String action, long value, String label) {
+        HitBuilders.EventBuilder eventBuilder = new HitBuilders.EventBuilder()
```
Ideas here: on collection open log the size of it, maybe the note count. If we set the language to a non-system language, log that so we can see if people actually use it, etc etc. We can log how long it takes to do imports or media checks or what-have-you.

mikehardy commented on AnkiDroid/src/main/java/com/ichi2/anki/dialogs/DeckPickerAnalyticsOptInDialog.java:
```diff
@@ -0,0 +1,39 @@
+                    AnkiDroidApp.getSharedPrefs(getContext()).edit()
+                            .putBoolean(UsageAnalytics.ANALYTICS_OPTIN_KEY, dialog.isPromptCheckBoxChecked())
+                            .apply();
+                    ((DeckPicker) getActivity()).dismissAllDialogFragments();
```
The change listener I install during init picks up on this pref edit with nothing else needed here

mikehardy commented on AnkiDroid/src/main/res/xml/preferences_analytics.xml:
```diff
@@ -0,0 +1,26 @@
+        android:key="analytics_opt_in"
+        android:summary="@string/analytics_summ"
+        android:title="@string/analytics_title" />
+</PreferenceScreen>
```
I should have deleted this - I'll make a commit that does so and we can squash later. It was from a previous attempt at a popup

timrae commented on AnkiDroid/build.gradle:
```diff
@@ -118,6 +121,7 @@ dependencies {
     implementation 'android.arch.persistence:db:1.1.1'
     implementation 'com.afollestad.material-dialogs:core:0.9.6.0'
     implementation 'com.getbase:floatingactionbutton:1.10.1'
+    implementation 'com.google.android.gms:play-services-analytics:16.0.3'
```
Doesn't this introduce a closed source dependency??

timrae replied:
Even if the library code itself is apache, if it finally uses google play services then it might not be GPL compatible, so we have to tread lightly

mikehardy replied:
Yeah, I don't want to hand-wave about it. Assume I'll re-do the core calls to use the ASL2 library linked above, but that dependency is fairly well isolated programmatically, so if you still have time to browse, especially the first commit where the work is really done, it shouldn't be time wasted

timrae replied:
It might take a while to get to through this

mikehardy replied:
totally understand. this feature out of all of them is exciting to me though - I'll try to be patient, plenty of other stuff to do ;-)

mikehardy commented:
I have implemented a polished version of the open-source google analytics library I want to use, in a way that works well on Android and I'm attempting to integrate it with upstream - cross-linking here as https://github.com/brsanthu/google-analytics-java/pull/59

If the google-analytics-java maintainer doesn't want to integrate I'll patch it in here. Either way you may generally assume that the little inner-class rats nest of OkHttp implementation will not be present in the final version, but the general structure will be identical to current.

TL;DR - no change in status at present - it needs review but please ignore the OkHttp inner-class bits for now.

mikehardy commented:
This is in it's final form now. I haven't gone crazy with tagging things that would answer hypotheses, this is just the basic implementation and tagging to see what the data looks like. A second pass would be focused tagging to answer questions and direct effort

timrae commented on AnkiDroid/build.gradle:
```diff
@@ -117,6 +123,9 @@ dependencies {
     implementation 'androidx.sqlite:sqlite:2.0.0-rc01'
     implementation 'com.afollestad.material-dialogs:core:0.9.6.0'
     implementation 'com.getbase:floatingactionbutton:1.10.1'
+    implementation 'net.mikehardy:google-analytics-java:2.0.1'
```
You're not getting any love upstream?

mikehardy replied:
No - it's a little disappointing. The library is simple and the upstream implementation is pretty good, but the author did a batch of work in 2014, then a batch of work in January, and otherwise mostly ignores issues and PRs. The protocol itself is basically one URL with a bunch of arguments so I figure what the heck...mostly maintenance mode anyway, just doesn't belong in our codebase. I'd been meaning to time travel back to 2008 and grab some maven central namespace anyway

mikehardy commented:
Anything stopping this one? Once the infrastructure is in I have a few questions I think could have interesting answers with proper analytics, and also we may need a sampling strategy to reduce hits but I won't know till see what the initial alpha data looks like...

timrae commented on AnkiDroid/src/main/java/com/ichi2/anki/analytics/AnkiDroidCrashReportDialog.java:
```diff
@@ -68,7 +66,7 @@ protected void init(Bundle savedInstanceState) {
         dialogBuilder.setView(buildCustomView(savedInstanceState));
 
-        mDialog = dialogBuilder.create();
+        AlertDialog mDialog = dialogBuilder.create();
```
Don't use `m` prefix for local variables

mikehardy commented on AnkiDroid/src/main/java/com/ichi2/anki/analytics/UsageAnalytics.java:
```diff
@@ -0,0 +1,227 @@
+     */
+    public static void sendAnalyticsScreenView(String screenName) {
+        Timber.d("sendAnalyticsScreenView(): %s", screenName);
+        sAnalytics.screenView().screenName(screenName).sendAsync();
```
Every hit is basically a GET with a bunch of URL params. If it's a batched hit (which we are using) then it's a POST, but really just a few lines of the GET style, POSTed at once. When you run it you can see them in the log. Here's a cut-paste of a batch from the library test suite:

```
[main] DEBUG org.apache.http.wire - http-outgoing-7 >> "POST /batch HTTP/1.1[\r][\n]"
[main] DEBUG org.apache.http.wire - http-outgoing-7 >> "Content-Length: 1560[\r][\n]"
[main] DEBUG org.apache.http.wire - http-outgoing-7 >> "Content-Type: application/x-www-form-urlencoded[\r][\n]"
[main] DEBUG org.apache.http.wire - http-outgoing-7 >> "Host: www.google-analytics.com[\r][\n]"
[main] DEBUG org.apache.http.wire - http-outgoing-7 >> "Connection: Keep-Alive[\r][\n]"
[main] DEBUG org.apache.http.wire - http-outgoing-7 >> "User-Agent: java/1.8.0_181-b13/Oracle Corporation/Java HotSpot(TM) 64-Bit Server VM/Linux/4.17.19-041719-generic/amd64[\r][\n]"
[main] DEBUG org.apache.http.wire - http-outgoing-7 >> "Accept-Encoding: gzip,deflate[\r][\n]"
[main] DEBUG org.apache.http.wire - http-outgoing-7 >> "[\r][\n]"
[main] DEBUG org.apache.http.wire - http-outgoing-7 >> "dt=Search&de=UTF-8&t=pageview&av=1.0.0&v=1&ul=en-US&dl=http%3A%2F%2Fwww.google.com&an=Junit+Test&tid=UA-612100-12&cid=fb5b686c-c097-490a-a1a5-8025cacbb8f1[\r][\n]"
[main] DEBUG org.apache.http.wire - http-outgoing-7 >> "dt=Search&de=UTF-8&t=pageview&av=1.0.0&v=1&ul=en-US&dl=http%3A%2F%2Fwww.google.com&an=Junit+Test&tid=UA-612100-12&cid=fb5b686c-c097-490a-a1a5-8025cacbb8f1[\r][\n]"
[main] DEBUG org.apache.http.wire - http-outgoing-7 >> "dt=Search&de=UTF-8&t=pageview&av=1.0.0&v=1&ul=en-US&dl=http%3A%2F%2Fwww.google.com&an=Junit+Test&tid=UA-612100-12&cid=fb5b686c-c097-490a-a1a5-8025cacbb8f1[\r][\n]"
[main] DEBUG org.apache.http.wire - http-outgoing-7 >> "dt=Search&de=UTF-8&t=pageview&av=1.0.0&v=1&ul=en-US&dl=http%3A%2F%2Fwww.google.com&an=Junit+Test&tid=UA-612100-12&cid=fb5b686c-c097-490a-a1a5-8025cacbb8f1[\r][\n]"
[main] DEBUG org.apache.http.wire - http-outgoing-7 >> "dt=Search&de=UTF-8&t=pageview&av=1.0.0&v=1&ul=en-US&dl=http%3A%2F%2Fwww.google.com&an=Junit+Test&tid=UA-612100-12&cid=fb5b686c-c097-490a-a1a5-8025cacbb8f1[\r][\n]"
[main] DEBUG org.apache.http.wire - http-outgoing-7 >> "dt=Search&de=UTF-8&t=pageview&av=1.0.0&v=1&ul=en-US&dl=http%3A%2F%2Fwww.google.com&an=Junit+Test&tid=UA-612100-12&cid=fb5b686c-c097-490a-a1a5-8025cacbb8f1[\r][\n]"
[main] DEBUG org.apache.http.wire - http-outgoing-7 >> "dt=Search&de=UTF-8&t=pageview&av=1.0.0&v=1&ul=en-US&dl=http%3A%2F%2Fwww.google.com&an=Junit+Test&tid=UA-612100-12&cid=fb5b686c-c097-490a-a1a5-8025cacbb8f1[\r][\n]"
[main] DEBUG org.apache.http.wire - http-outgoing-7 >> "dt=Search&de=UTF-8&t=pageview&av=1.0.0&v=1&ul=en-US&dl=http%3A%2F%2Fwww.google.com&an=Junit+Test&tid=UA-612100-12&cid=fb5b686c-c097-490a-a1a5-8025cacbb8f1[\r][\n]"
[main] DEBUG org.apache.http.wire - http-outgoing-7 >> "dt=Search&de=UTF-8&t=pageview&av=1.0.0&v=1&ul=en-US&dl=http%3A%2F%2Fwww.google.com&an=Junit+Test&tid=UA-612100-12&cid=fb5b686c-c097-490a-a1a5-8025cacbb8f1[\r][\n]"
[main] DEBUG org.apache.http.wire - http-outgoing-7 >> "dt=Search&de=UTF-8&t=pageview&av=1.0.0&v=1&ul=en-US&dl=http%3A%2F%2Fwww.google.com&an=Junit+Test&tid=UA-612100-12&cid=fb5b686c-c097-490a-a1a5-8025cacbb8f1[\r][\n]"
[main] DEBUG org.apache.http.wire - http-outgoing-7 << "HTTP/1.1 200 OK[\r][\n]"
[main] DEBUG org.apache.http.wire - http-outgoing-7 << "Date: Sat, 06 Oct 2018 02:41:04 GMT[\r][\n]"
[main] DEBUG org.apache.http.wire - http-outgoing-7 << "Pragma: no-cache[\r][\n]"
[main] DEBUG org.apache.http.wire - http-outgoing-7 << "Expires: Fri, 01 Jan 1990 00:00:00 GMT[\r][\n]"
[main] DEBUG org.apache.http.wire - http-outgoing-7 << "Cache-Control: no-cache, no-store, must-revalidate[\r][\n]"
[main] DEBUG org.apache.http.wire - http-outgoing-7 << "Last-Modified: Sun, 17 May 1998 03:00:00 GMT[\r][\n]"
[main] DEBUG org.apache.http.wire - http-outgoing-7 << "X-Content-Type-Options: nosniff[\r][\n]"
[main] DEBUG org.apache.http.wire - http-outgoing-7 << "Content-Type: image/gif[\r][\n]"
[main] DEBUG org.apache.http.wire - http-outgoing-7 << "Server: Golfe2[\r][\n]"
[main] DEBUG org.apache.http.wire - http-outgoing-7 << "Content-Length: 35[\r][\n]"
[main] DEBUG org.apache.http.wire - http-outgoing-7 << "Alt-Svc: quic=":443"; ma=2592000; v="44,43,39,35"[\r][\n]"
[main] DEBUG org.apache.http.wire - http-outgoing-7 << "[\r][\n]"
[main] DEBUG org.apache.http.wire - http-outgoing-7 << "GIF89a[0x1][0x0][0x1][0x0][0x80][0xff][0x0][0xff][0xff][0xff][0x0][0x0][0x0],[0x0][0x0][0x0][0x0][0x1][0x0][0x1][0x0][0x0][0x2][0x2]D[0x1][0x0];"

```

cid is client id, that's the only thing that ties the user session together. The IP address obviously goes out with the hit but we ask google to anonymous them and the clobber the last octet of the IP on their end. Other than that it's the 1.0 protocol version, our property id (UA-xxxxxx), the event type (almost always screen for us, sometimes exception, event, and I'd like to do some timings, but none of the commerce obviously). And the types are loose - we can use whatever ASCII string we want as a tag and it gets aggregated, in that way we can make synthetic names for preferences subscreens for instance.

mikehardy commented on AnkiDroid/src/main/java/com/ichi2/anki/analytics/UsageAnalytics.java:
```diff
@@ -0,0 +1,227 @@
+    public static void sendAnalyticsException(@NonNull Throwable t, boolean fatal)
+    {
+        // Get at the root cause if possible
+        Throwable root = t;
```
must be a remnant - that's no good, I'll go over it finely

mikehardy replied:
not a remnant - infinite logic bomb. I want to keep following exception chain and send the most internally one, the loop as written kept cycling on the top two exceptions. Thank you for the review!

mikehardy commented on AnkiDroid/src/main/java/com/ichi2/anki/analytics/AnalyticsDialogFragment.java:
```diff
@@ -0,0 +1,11 @@
+
+public abstract class AnalyticsDialogFragment extends DialogFragment {
+    @Override
+    public void onResume() {
```
I had the idea I considered a loss of interaction on a screen (onPause()) followed by resumption a fresh screen hit. If it's just onCreate the chain of taps didn't seem quite right to me? This might be something we learn with data and change our minds on (from either starting point) though. You prefer onCreate()?

mikehardy commented on AnkiDroid/src/main/res/values/10-preferences.xml:
```diff
@@ -237,4 +237,9 @@
+    <!-- analytics options -->
+    <string name="analytics_dialog_title">Help make AnkiDroid better!</string>
+    <string name="analytics_title">Share feature usage</string>
+    <string name="analytics_summ">You can contribute to AnkiDroid by helping the development team see which features people use.</string>
```
If you want to change onResume() to onCreate() I'll alter one more thing, otherwise I've implemented your stance on all comments and rebased-pushed. I'm done of the day so no more yammering from over here. Cheers

mikehardy commented:
If you want to change onResume() to onCreate() I'll alter one more thing, otherwise I've implemented your stance on all comments and rebased-pushed. I'm done for the day so no more yammering from over here. Cheers

timrae commented on AnkiDroid/src/main/java/com/ichi2/anki/analytics/UsageAnalytics.java:
```diff
@@ -205,10 +205,10 @@ public static void sendAnalyticsEvent(@NonNull String category, @NonNull String
-            root = t.getCause();
+            root = root.getCause();
         }
         sendAnalyticsException(t.getClass().getSimpleName(), fatal);
```
`root` variable is still unused, isn't it...?

mikehardy replied:
Yes...so, got tired of sending head-smackers in on this PR in this area so actually wrote the test then made sure the implementation was correct...https://github.com/ankidroid/Anki-Android/pull/4969/commits/c44e8b2c181c721a582698ec16d104e04caa78da#diff-d680d134e50a11500ad20bd88363bcf3R82

mikehardy commented:
@timrae maybe actually ready to go now that I'm not screwing up exception root cause finding...

timrae commented:
What are the limits for free usage again? I'm guessing we'll saturate those pretty quickly when we roll this out to production... Currently it seems we're mainly looking at Activity usage patterns. That will generate a lot of traffic and I'm not totally sure how useful or not that is. I guess I'm more interested in usage of specific features. Is there a way at the moment to see stats on our users settings for example?

mikehardy commented:
I was counting on a little bit of information from the first few alpha builds (and maybe a beta) to see if I needed to implement the sampling strategy - but the limits are actually pretty high:

https://github.com/ankidroid/Anki-Android/pull/4969/commits/8837bb1b9de086889fb5a92566bd688a484b86dd#diff-59d7cc4c9cf92f44d49234387d7dc27bR5

And note that Reviewer (for instance) doesn't send hits for every review, only if it toggles between onPause()/onResume() (which it does not during normal review)

I will definitely implement a sampling strategy if reasonable extrapolation predicts we blow past the limits, while at the same time, all blowing past the limits does is get you a contact from google if it's the whole 10MM / month limit, or they drop hits for the session or day if it's a local peak - https://developers.google.com/analytics/devguides/collection/analyticsjs/limits-quotas

timrae commented:
> Is there a way at the moment to see stats on our users settings for example?

This is quite interesting to me

timrae commented:
Also, any idea on potential battery life degradation from the increased web traffic on this?

mikehardy commented:
1.  Re: limits - a fair point, I'll begin work on sampling strategy but I'd still like to merge prior so we can start seeing what it looks like for the next item:
1.  Users settings - I don't understand what you want, but if you want to send an event that contains what they have turned on or off, we could certainly have a sort of "preferences" hit with whatever you wanted to know?
1.  Battery life degradation - unsure. I know we'll only be sending data when the screen's on, the connection is persistent (thus efficient) and the data is small. There shouldn't be that many hits in a normal review session...should be fine? But I don't know honestly. I don't want a race to the bottom but I feel like this is what all apps do these days so it shouldn't be abnormal. If we sent a hit for each review, then yeah...

As for safety on the limits just in case sampling strategy doesn't work, the simplest sample is literally modulo on a random number for any given session to tune it down, or we can just disable - so I'm not worried about being able to fix it but I do want to have time for both of us to see what data looks like and instrument interesting things before it goes large-scale

timrae commented:
>  I don't understand what you want, but if you want to send an event that contains what they have turned on or off, we could certainly have a sort of "preferences" hit with whatever you wanted to know?

I want to know that and also somehow "x percent of users have this setting enabled"

timrae commented:
Anyway that can be a feature request, I think this is ready to merge with the caveat that the other stuff will need to be implemented before putting into prod

mikehardy commented:
Okay - I'm about to set down the RAID can I've been spraying the bugs with and do sampling strategy in the library so it won't be long. "x percent of users have this feature" I'll think how to tag it but it must be possible. I can think of around 5 or 6 data points now worth tagging - it should be really useful. Or, if totally useless it'll be easy to rip out :)

timrae commented:
...and now we have a commit called "Respond to review - will re-shape after review for final PR" in the commit history... :/

mikehardy commented:
Aw crap. Seriously, crap. I'm sorry. It's probably just us two, I could force push if you call it quickly.

timrae commented:
Yeah ok, but please take a backup before pushing and double check you are on the correct branch and pushing to the correct place

mikehardy commented:
Okay, that gave me a nice nervous sweat. I have a git --clone mirror. I read a couple different docs to make sure, I checked I'd fetched and was otherwise up to date. I git reset --soft HEAD~N on my N commits that should have been 2 nice ones, I re-shaped them, commited and did the push upstream +master after resetting the URL. I'm going to reset the push URL to a NOP and not do this again...re-pull and clean your tree and we should be good

timrae commented:
Ok seems to be fine now. In the future, please keep in mind that if I approve something without merging it, it means I believe it's not quite ready to merge as-is, or at least there is a reason why I didn't merge it, so please ask if it's not obvious.

mikehardy commented:
Yes, obvious in hind-sight esp since it was in my own commit message, and I'll have a difficult time forgetting after this. I was just starting to feel good at git but "git push upstream +master" was a fearsome thing to type

mikehardy commented:
@timrae As we approach release, I investigated the current analytics hit rate to see if we'd be okay vs the back-end limits.

- super safe on per session limit
- super safe on per user limit
- optimally safe on global hit count limit! we will consume 85.2% of the limit it appears, at the end of 2 years, which is a good trade-off between capturing what we can, but not blowing up?

details for future explorers:

Current numbers:
open beta participants: 7602
current installs of beta: 3190
current 28-day users in analytics: 981 (not all will be sampled every day)
current 1-day users in analytics: 149 (reasonable bet on daily # of samples vs population)
current 1-day total hit count avg from 7-day count: 2341
current 1-day total hits-per-opted-in-user: 15.7
current per-session hit count avg: 12 (vs 500 per session limit)
current per-user session count avg: about 2
current per-user hit count avg: about 24 then

current active user count 1,170,000
current growth rate about 300,000 a year

goal for "time until needs tuning again" 2 years? so plan for 1,750,000 million maybe?

current opt-in rate: 31%
expected opt-in users from goal end user count: 542,500
expected monthly hit count from expected goal end user count: 8,523,439

Hit count Limits: 10MM/month globally, 200K/user && 500/session, 1 ever 2s + 60 in a session

timrae commented:
Sounds good!
